### ISSUE TITLE
linter: typescript/no-unnecessary-condition false positive on ??= with noUncheckedIndexedAccess

### ISSUE THREAD

**Issue by mdierolf** [oxc-project/oxc#24772]
# `typescript/no-unnecessary-condition` false positive on `??=` with `noUncheckedIndexedAccess`

**Note: this bug was noted by a human; the reproduction was created with the assistance of AI**

Self contained repo below, unzip and run oxlint & tsc to see the failing and not failing case:
[oxlint-optional-equals-bug.zip](https://github.com/user-attachments/files/30240122/oxlint-optional-equals-bug.zip)

## Summary

With type-aware linting enabled, `typescript/no-unnecessary-condition` incorrectly flags nullish coalescing assignment (`??=`) on a `Record` index access as:

```text
Unnecessary optional chain on a non-nullish value
```

This is wrong for two reasons:

1. **There is no optional chain (`?.`)** on the flagged line — the operator is `??=`.
2. **Under `noUncheckedIndexedAccess`, the LHS is correctly `T | undefined`**, so `??=` is necessary. TypeScript itself accepts the code and types the access as possibly undefined.

Semantically equivalent forms (`if (x === undefined) x = {}` and `x ?? {}`) are **not** flagged.

## Environment

| Package | Version |
|---------|---------|
| `oxlint` | 1.74.0 |
| `oxlint-tsgolint` | 0.24.0 |
| `typescript` | 5.8.3 (project; type-aware analysis via tsgolint) |
| OS | macOS (also expected on other platforms) |

## Minimal reproduction

This directory is a self-contained repro.

```bash
npm install   # or: bun install / pnpm install
npm run repro # or: npx oxlint -f stylish src/repro.ts
```

### Config

**`tsconfig.json`** (relevant bits):

```json
{
  "compilerOptions": {
    "strict": true,
    "noUncheckedIndexedAccess": true
  }
}
```

**`.oxlintrc.json`** (relevant bits):

```json
{
  "options": {
    "typeAware": true,
    "typeCheck": true
  },
  "rules": {
    "typescript/no-unnecessary-condition": "error"
  }
}
```

### Code (`src/repro.ts`)

```ts
type ProviderOptions = Record<string, Record<string, unknown>>;

export function withNullishAssignment(opts: ProviderOptions): Record<string, unknown> {
  // ❌ FALSE POSITIVE
  const perplexity = opts.perplexity ??= {};
  return perplexity;
}

export function withIfUndefined(opts: ProviderOptions): Record<string, unknown> {
  // ✅ not flagged (equivalent mutation)
  if (opts.perplexity === undefined) {
    opts.perplexity = {};
  }
  return opts.perplexity;
}

export function withNullishCoalesce(opts: ProviderOptions): Record<string, unknown> {
  // ✅ not flagged
  return opts.perplexity ?? {};
}
```

## Actual behavior

```text
$ npm run repro

file:///…/src/repro.ts
  16:22  error  Unnecessary optional chain on a non-nullish value.  typescript(no-unnecessary-condition)

✖ 1 problem (1 error, 0 warnings)
```

Only the `??=` form is reported. The `if === undefined` and `??` forms in the same file are clean. `tsc --noEmit` exits 0 under the same `tsconfig.json`.

## Expected behavior

No diagnostic on:

```ts
const perplexity = opts.perplexity ??= {};
```

With `noUncheckedIndexedAccess`, `opts.perplexity` has type `Record<string, unknown> | undefined`. The assignment is a valid, necessary nullish check + write.

If a diagnostic is ever warranted for a truly non-nullish LHS, the message should describe **nullish coalescing assignment**, not “optional chain”.

## Why this matters

`??=` is the idiomatic way to ensure a nested options bag exists and then mutate it:

```ts
const bag = request.providerOptions.perplexity ??= {};
bag.someFlag = true;
```

The suggested workaround (`if` + assign) is more verbose and identical in intent. Bare `??` is **not** equivalent because it does not write back to the object when the key is missing.

## Notes

- Reproduced with type-aware mode (`options.typeAware` / `typeCheck` + `oxlint-tsgolint`).
- Classic (non-type-aware) oxlint does not apply this TypeScript rule the same way; type-aware mode is required.
- TypeScript compiles the file without error under the same `tsconfig.json`.

## Suggested fix direction

- Treat `??=` / `||=` / `&&=` using the same nullability analysis as `??` / `||` / `&&` and as explicit `=== undefined` checks.
- When reporting on `??=`, use a message that names nullish coalescing assignment rather than optional chaining.
- Ensure `noUncheckedIndexedAccess` is honored for property/index access on `Record<string, T>` (and index signatures) when evaluating whether the LHS of `??=` can be nullish.
